src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.Core.cs

```diff
@@ -69,19 +65,34 @@ private object DeserializeObject(int typeIndex)
 
             if (Volatile.Read(ref _binaryFormatter) is null)
             {
-                if (!InitializeBinaryFormatterLocal())
+                try
                 {
-                    // Trimming took away the BinaryFormatter implementation and we can't call into it.
-                    // We'll throw an exception with the same text that BinaryFormatter would have thrown
-                    // had we been able to call into it. Keep this resource string in sync with the same
-                    // resource from the Formatters assembly.
-                    throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed);
+                    if (!InitializeBinaryFormatterLocal())
+                    {
+                        // Trimming took away the BinaryFormatter implementation and we can't call into it.
+                        // We'll throw an exception with the same text that BinaryFormatter would have thrown
+                        // had we been able to call into it. Keep this resource string in sync with the same
+                        // resource from the Formatters assembly.
+                        throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed);
+                    }
+                }
+                catch (Exception ex) when (ex is TypeLoadException or MissingMethodException)
+                {
+                    throw new NotSupportedException(SR.BinaryFormatter_SerializationDisallowed, ex);
                 }
             }
 
             Type type = FindType(typeIndex);
 
-            object graph = s_deserializeMethod!(_binaryFormatter, _store.BaseStream);
+            object graph;
+            try
+            {
+                graph = s_deserializeMethod!(_binaryFormatter!, _store.BaseStream);
```

Comment:
This can call the UnsafeAccessor directly, without try/catch, s_deserializeMethod cache, etc.